File: apps/cargo-tracker/src/main/java/com/example/cargotracker/billing/domain/model/repository/InvoiceRepository.java
package com.example.cargotracker.billing.domain.model.repository;

import com.example.cargotracker.billing.domain.model.aggregates.Invoice;
import com.example.cargotracker.billing.domain.model.aggregates.InvoiceId;

import java.util.List;
import java.util.Optional;

public interface InvoiceRepository {
    void save(Invoice invoice);
    Optional<Invoice> findById(InvoiceId id);
    Optional<Invoice> findByFreightChargeId(String freightChargeId);
    List<Invoice> findAll();
}

File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/billing/domain/model/repositories/InvoiceRepository.scala
package cargotracker.billing.domain.model.repositories

import cargotracker.billing.domain.model.aggregates.Invoice
import cargotracker.billing.domain.model.valueobjects.{BillingBookingId, InvoiceId}

/** Billing Context のリポジトリポート（ヘキサゴナル / US21）。 */
trait InvoiceRepository:
  def nextInvoiceId(): InvoiceId
  def findById(id: InvoiceId): Option[Invoice]
  def findByBookingId(bookingId: BillingBookingId): Option[Invoice]
  def findAll(): Seq[Invoice]
  def save(invoice: Invoice): Unit
